File: sdk/conformance/harness/python/conformance/_report.py
"""JUnit XML emitter for the Python harness.

Tight mirror of the Go harness's report shape — downstream CI tooling
(step 8's aggregator) consumes both streams under the same schema.
"""

from __future__ import annotations

import os
import xml.etree.ElementTree as ET
from dataclasses import dataclass, field
from typing import Optional


@dataclass
class TestCase:
    name: str
    class_name: str
    time_sec: float
    failure_message: Optional[str] = None
    failure_body: str = ""


@dataclass
class TestSuite:
    name: str
    cases: list[TestCase] = field(default_factory=list)

    @property
    def tests(self) -> int:
        return len(self.cases)

    @property
    def failures(self) -> int:
        return sum(1 for c in self.cases if c.failure_message)

    def to_xml(self) -> bytes:
        root = ET.Element(
            "testsuite",
            attrib={
                "name": self.name,
                "tests": str(self.tests),
                "failures": str(self.failures),
            },
        )
        for c in self.cases:
            tc = ET.SubElement(
                root,
                "testcase",
                attrib={
                    "name": c.name,
                    "classname": c.class_name,
                    "time": f"{c.time_sec:.7f}",
                },
            )
            if c.failure_message:
                fail = ET.SubElement(
                    tc,
                    "failure",
                    attrib={"message": c.failure_message, "type": "AssertionError"},
                )
                fail.text = c.failure_body
        ET.indent(root, space="  ")
        return b'<?xml version="1.0" encoding="UTF-8"?>\n' + ET.tostring(root, encoding="utf-8")


def write_report(path: str, suite: TestSuite) -> None:
    os.makedirs(os.path.dirname(path) or ".", exist_ok=True)
    with open(path, "wb") as f:
        f.write(suite.to_xml())

File: sdk/conformance/harness/typescript/src/report.mjs
// JUnit XML emitter for the TypeScript harness. Same schema as the
// Go + Python harnesses so step-7's aggregator consumes any of them.

import { promises as fs } from 'node:fs';
import path from 'node:path';

export class TestCase {
  constructor(name, className, timeSec) {
    this.name = name;
    this.className = className;
    this.timeSec = timeSec;
    this.failureMessage = null;
    this.failureBody = '';
  }
}

export class TestSuite {
  constructor(name) {
    this.name = name;
    this.cases = [];
  }

  get tests() { return this.cases.length; }

  get failures() {
    return this.cases.filter((c) => c.failureMessage).length;
  }

  toXML() {
    const esc = (s) =>
      String(s)
        .replaceAll('&', '&amp;')
        .replaceAll('<', '&lt;')
        .replaceAll('>', '&gt;')
        .replaceAll('"', '&quot;')
        .replaceAll("'", '&apos;');
    const lines = [
      '<?xml version="1.0" encoding="UTF-8"?>',
      `<testsuite name="${esc(this.name)}" tests="${this.tests}" failures="${this.failures}">`,
    ];
    for (const c of this.cases) {
      const head = `  <testcase name="${esc(c.name)}" classname="${esc(c.className)}" time="${c.timeSec.toFixed(7)}"`;
      if (c.failureMessage) {
        lines.push(`${head}>`);
        lines.push(`    <failure message="${esc(c.failureMessage)}" type="AssertionError">${esc(c.failureBody)}</failure>`);
        lines.push('  </testcase>');
      } else {
        lines.push(`${head}></testcase>`);
      }
    }
    lines.push('</testsuite>');
    return lines.join('\n') + '\n';
  }
}

export async function writeReport(outPath, suite) {
  const dir = path.dirname(outPath);
  await fs.mkdir(dir, { recursive: true });
  await fs.writeFile(outPath, suite.toXML(), 'utf8');
}
